[SPARK-17378] [BUILD] Upgrade snappy-java to 1.1.2.6

a-roberts opened the pull request:
## What changes were proposed in this pull request?

Upgrades the Snappy version to 1.1.2.6 from 1.1.2.4, release notes: https://github.com/xerial/snappy-java/blob/master/Milestone.md mention "Fix a bug in SnappyInputStream when reading compressed data that happened to have the same first byte with the stream magic header (#142)"
## How was this patch tested?

Existing unit tests using the latest IBM Java 8 on Intel, Power and Z architectures (little and big-endian)

a-roberts committed:
Update Snappy to 1.1.2.6

Pom change

a-roberts committed:
Update spark-deps-hadoop-2.2

a-roberts committed:
Update spark-deps-hadoop-2.3

a-roberts committed:
Update spark-deps-hadoop-2.4

a-roberts committed:
Update spark-deps-hadoop-2.6

a-roberts committed:
Update spark-deps-hadoop-2.7

SparkQA commented:
**[Test build #64932 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/64932/consoleFull)** for PR 14958 at commit [`3795997`](https://github.com/apache/spark/commit/3795997a116c868343930ac9a91afb861c82705c).

srowen commented:
Looks OK for master/2.0, and even 1.6 as it's already on 1.1.2.1, and changes from 1.1.2.1 to 1.1.2.4 look like strict fixes/improvements. https://github.com/xerial/snappy-java/blob/master/Milestone.md

a-roberts commented:
Yep, builds in progress on our farm against the latest 1.6 code using 1.1.2.6 so will improve the PR should there be no problems

jaceklaskowski commented:
Would be nice to have a test in Spark to see if the changes to snappy made any sense to Spark (not that I'm against -- just as a safety measure that it did improve things if possible).

a-roberts commented:
Good point, https://github.com/xerial/snappy-java/commit/60cc0c2e1d1a76ae2981d0572a5164fcfdfba5f1 shows the changes that were made, so we would have a checkMagicHeader equivalent added to the existing Spark tests, I'll look into it

SparkQA commented:
**[Test build #64932 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/64932/consoleFull)** for PR 14958 at commit [`3795997`](https://github.com/apache/spark/commit/3795997a116c868343930ac9a91afb861c82705c).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/64932/
Test PASSed.

srowen commented:
I don't know that the bug that was fixed has manifested yet in Spark; certainly not in the unit tests. It seems like something that could affect some execution out in the wild. It seems reasonable to take the fix as I don't see a downside to it. I don't know if we need to reproduce this same bug; snappy tests have done that.

jaceklaskowski commented:
Fair enough. It was just an idea that would make the upgrade even more super-needed. The issues that snappy has fixed could manifest in various ways in Spark and while doing so could uncover others. Yes, again, it could be a project on its own. Sorry for the noise.

markhamstra commented:
@srowen The bug does manifest in Spark.  I've been running this upgrade in production for a few weeks with no issues.  This should be merged IMO.

a-roberts commented:
I'm almost done with adding in the test case anyway, involved a quick rewrite to use Scala code, had fun rewriting the for loop for the test into a while, I _think_ it would manifest itself if your very first byte happened to be -126 and you're using Snappy.

At the very least I brushed up on my Scala skills and I think it's good practice to add a test with an implementation change so I'll update the PR once it's done. If Snappy were to change (e.g. 1.1.3 were to be released but the magic header logic changes) we'd at least be another project that picks up the problem.

srowen commented:
Yeah I'm cool to merge this now, but if you have a test, OK. Is it testing Spark or just testing Snappy though?

a-roberts committed:
Add test for snappy-java handling of magic header

Probably needs TLC with the while loop, this is based on https://github.com/xerial/snappy-java/blob/60cc0c2e1d1a76ae2981d0572a5164fcfdfba5f1/src/test/java/org/xerial/snappy/SnappyInputStreamTest.java but outside of the snappy package so we can't use MAGIC_HEADER[0].

a-roberts commented:
Straight up rewrite so it's just testing that the Snappy version we include happens to have the magic header handling in, kept it simple

a-roberts commented on core/src/test/scala/org/apache/spark/io/CompressionCodecSuite.scala:
```diff
@@ -130,4 +130,58 @@ class CompressionCodecSuite extends SparkFunSuite {
+      var buf = new Array[Byte](4096)
+
+      var readBytes = 0
+      while (readBytes != -1) {
```
This loop makes me weep, it's an attempt @ rewriting the following but in Scala

 for (int readBytes = 0; (readBytes = input.read(buf)) != -1; ) {

SparkQA commented:
**[Test build #65000 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/65000/consoleFull)** for PR 14958 at commit [`accc88f`](https://github.com/apache/spark/commit/accc88f8c73864d50b5588f43c53c11323052d2e).

SparkQA commented:
**[Test build #65000 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/65000/consoleFull)** for PR 14958 at commit [`accc88f`](https://github.com/apache/spark/commit/accc88f8c73864d50b5588f43c53c11323052d2e).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/65000/
Test FAILed.

srowen commented on core/src/test/scala/org/apache/spark/io/CompressionCodecSuite.scala:
```diff
@@ -130,4 +130,58 @@ class CompressionCodecSuite extends SparkFunSuite {
+  test("SPARK 17378: snappy-java should handle magic header when reading stream") {
+    val b = new ByteArrayOutputStream()
+    // Write uncompressed length beginning with -126 (the same with magicheader[0])
+    b.write(-126) // Can't access magic header[0] as it isn't public, so access this way
```
Hm, this seems to tie this test to this internal detail of Snappy though. Spark itself doesn't really need to assert this detail in a test. 

I feel like this test is just testing snappy, which snappy can test. I could see testing a case at the level of Spark that triggers this bug and verifies it's fixed.

a-roberts replied:
Yeah I agree, so how about we just revert the test case commit here and merge the 1.1.2.6 change itself as folks want it, and then in a later PR add an extra test for robustness if we want to.

a-roberts committed:
Revert test addition in CompressionCodecSuite

Was a duplicate of the Snappy test instead of exercising the problem from Spark, something to think about later

SparkQA commented:
**[Test build #65001 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/65001/consoleFull)** for PR 14958 at commit [`022cad7`](https://github.com/apache/spark/commit/022cad7b69f75b2e358fdc1b0f01224c3c067424).

SparkQA commented:
**[Test build #65001 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/65001/consoleFull)** for PR 14958 at commit [`022cad7`](https://github.com/apache/spark/commit/022cad7b69f75b2e358fdc1b0f01224c3c067424).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/65001/
Test PASSed.

srowen commented:
Merged to master/2.0/1.6

yhuai commented:
```
Using `mvn` from path: /home/jenkins/workspace/spark-branch-1.6-lint/build/apache-maven-3.3.9/bin/mvn
Spark's published dependencies DO NOT MATCH the manifest file (dev/spark-deps).
To update the manifest file, run './dev/test-dependencies.sh --replace-manifest'.
diff --git a/dev/deps/spark-deps-hadoop-1 b/dev/pr-deps/spark-deps-hadoop-1
index dd5a6dc..a97b10c 100644
--- a/dev/deps/spark-deps-hadoop-1
+++ b/dev/pr-deps/spark-deps-hadoop-1
@@ -143,7 +143,7 @@ servlet-api-2.5.jar
 slf4j-api-1.7.10.jar
 slf4j-log4j12-1.7.10.jar
 snappy-0.2.jar
-snappy-java-1.1.2.1.jar
+snappy-java-1.1.2.6.jar
 spire-macros_2.10-0.7.4.jar
 spire_2.10-0.7.4.jar
 stax-api-1.0.1.jar
Using `mvn` from path: /home/jenkins/workspace/spark-branch-1.6-lint/build/apache-maven-3.3.9/bin/mvn
Build step 'Execute shell' marked build as failure
Finished: FAILURE
```

Can you take a look at 1.6 build (https://amplab.cs.berkeley.edu/jenkins/view/Spark%20QA%20Compile/job/spark-branch-1.6-lint/262/console)? Seems the 1.6 build is broken by this pr.

srowen commented:
@yhuai ack OK I'll fix that. I must have managed to miss something in the deps file when I resolved conflicts on the cherry pick. It can't be complex to touch up though.
